File: liveCOMpiler.py
#!/usr/bin/env python3
"""liveCOMpiler.py"""
import sys, os, hashlib, time, subprocess, threading

POLL_INTERVAL = 0.5

def clear():
    os.system("cls" if os.name == "nt" else "clear")

def md5_of_file(path: str) -> str:
    h = hashlib.md5()
    with open(path, "rb") as f:
        for chunk in iter(lambda: f.read(8192), b""):
            h.update(chunk)
    return h.hexdigest()

def compile_and_run(src: str, binary: str, proc_holder: list) -> None:
    """Compiles src, then runs the binary. proc_holder[0] holds the Popen handle."""
    clear()
    result = subprocess.run(
        ["gcc", "-o", binary, src, "-lm"],
        capture_output=True, text=True,
    )
    if result.returncode != 0:
        print(f"error: {os.path.basename(src)}\n")
        print(result.stderr.strip())
        return

    proc = subprocess.Popen([binary])
    proc_holder[0] = proc        # expose handle so watcher can kill it
    proc.wait()                  # block *this thread* (not the watcher loop)
    proc_holder[0] = None
    print(f"\n— exited · watching {os.path.basename(src)} —")

def kill_current(proc_holder: list, thread_holder: list) -> None:
    """Kill the running child process and wait for its thread to finish."""
    proc = proc_holder[0]
    if proc and proc.poll() is None:   # still running
        proc.kill()
        proc.wait()
    t = thread_holder[0]
    if t and t.is_alive():
        t.join()

def watch(src: str) -> None:
    if not os.path.isfile(src):
        print(f"error: '{src}' not found.")
        sys.exit(1)

    src_dir = os.path.dirname(os.path.abspath(src))
    binary  = os.path.join(src_dir, ".livec_out")

    proc_holder   = [None]   # proc_holder[0]   = current Popen (or None)
    thread_holder = [None]   # thread_holder[0] = current Thread (or None)
    last_hash     = ""

    try:
        while True:
            try:
                current_hash = md5_of_file(src)
            except OSError:
                time.sleep(POLL_INTERVAL)
                continue

            if current_hash != last_hash:
                last_hash = current_hash

                # ── interrupt whatever is currently running ──────────────
                kill_current(proc_holder, thread_holder)

                # ── spin up a fresh compile-and-run in its own thread ────
                t = threading.Thread(
                    target=compile_and_run,
                    args=(src, binary, proc_holder),
                    daemon=True,
                )
                thread_holder[0] = t
                t.start()

            time.sleep(POLL_INTERVAL)

    except KeyboardInterrupt:
        kill_current(proc_holder, thread_holder)
        if os.path.exists(binary):
            os.remove(binary)
        sys.exit(0)

if __name__ == "__main__":
    if len(sys.argv) != 2:
        print("usage: python3 liveCOMpiler.py <file.c>")
        sys.exit(1)
    watch(sys.argv[1])

File: liveCOMpiler.c
/*
 * liveCOMpiler.c
 * Compile & run a C file, restarting automatically on save.
 *
 * Build:  gcc -o livecompiler liveCOMpiler.c -lpthread -lm
 * Usage:  ./livecompiler <file.c>
 */

#define _POSIX_C_SOURCE 200809L
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>
#include <unistd.h>
#include <signal.h>
#include <pthread.h>
#include <sys/stat.h>
#include <sys/wait.h>
#include <sys/types.h>
#include <limits.h>
#include <errno.h>

#define POLL_MS 500

/* ── shared state (guarded by mutex) ──────────────────────────────────── */
static pthread_mutex_t  g_lock  = PTHREAD_MUTEX_INITIALIZER;
static pthread_t        g_tid;          /* current worker thread          */
static int              g_tid_valid = 0;
static volatile pid_t   g_child_pid = -1; /* PID of running binary        */

/* ── paths ────────────────────────────────────────────────────────────── */
static char g_src[PATH_MAX];
static char g_binary[PATH_MAX];

/* ─────────────────────────────────────────────────────────────────────── */

static void clear(void)
{
    system("clear");
}

/* Simple file fingerprint using mtime + size (no MD5 dep needed). */
static uint64_t file_stamp(const char *path)
{
    struct stat st;
    if (stat(path, &st) != 0) return 0;
    return ((uint64_t)st.st_mtime << 32) | (uint64_t)st.st_size;
}

/* Kill the child binary if it is still running. */
static void kill_child(void)
{
    pid_t pid = g_child_pid;
    if (pid > 0) {
        kill(pid, SIGKILL);
        /* reap it so we don't leave a zombie */
        waitpid(pid, NULL, 0);
        g_child_pid = -1;
    }
}

/* ── worker thread: compile then run ─────────────────────────────────── */
static void *worker(void *arg)
{
    (void)arg;

    clear();

    /* ── compile ── */
    pid_t cpid = fork();
    if (cpid == 0) {
        /* child: exec gcc */
        execlp("gcc", "gcc", "-o", g_binary, g_src, "-lm", (char *)NULL);
        _exit(127);
    }
    if (cpid < 0) { perror("fork"); return NULL; }

    int status = 0;
    waitpid(cpid, &status, 0);
    if (!WIFEXITED(status) || WEXITSTATUS(status) != 0) {
        fprintf(stderr, "error: compilation failed\n");
        return NULL;
    }

    /* ── run ── */
    pid_t rpid = fork();
    if (rpid == 0) {
        execlp(g_binary, g_binary, (char *)NULL);
        _exit(127);
    }
    if (rpid < 0) { perror("fork"); return NULL; }

    /* publish PID so the watcher can kill it */
    pthread_mutex_lock(&g_lock);
    g_child_pid = rpid;
    pthread_mutex_unlock(&g_lock);

    waitpid(rpid, NULL, 0);

    pthread_mutex_lock(&g_lock);
    g_child_pid = -1;
    pthread_mutex_unlock(&g_lock);

    printf("\n— exited · watching %s —\n", g_src);
    fflush(stdout);
    return NULL;
}

/* Stop the current worker thread + child process. */
static void stop_current(void)
{
    pthread_mutex_lock(&g_lock);
    kill_child();
    int valid = g_tid_valid;
    pthread_mutex_unlock(&g_lock);

    if (valid) {
        pthread_cancel(g_tid);
        pthread_join(g_tid, NULL);
        g_tid_valid = 0;
    }
}

/* Spawn a fresh worker thread. */
static void start_worker(void)
{
    pthread_attr_t attr;
    pthread_attr_init(&attr);
    pthread_attr_setdetachstate(&attr, PTHREAD_CREATE_JOINABLE);

    if (pthread_create(&g_tid, &attr, worker, NULL) != 0) {
        perror("pthread_create");
    } else {
        g_tid_valid = 1;
    }
    pthread_attr_destroy(&attr);
}

/* ── clean exit on Ctrl-C ─────────────────────────────────────────────── */
static volatile sig_atomic_t g_quit = 0;
static void on_sigint(int s) { (void)s; g_quit = 1; }

/* ── watch loop ───────────────────────────────────────────────────────── */
static void watch(void)
{
    uint64_t last_stamp = 0;

    while (!g_quit) {
        uint64_t cur = file_stamp(g_src);

        if (cur != 0 && cur != last_stamp) {
            last_stamp = cur;
            stop_current();
            start_worker();
        }

        struct timespec ts = { 0, POLL_MS * 1000000L };
        nanosleep(&ts, NULL);
    }

    /* cleanup */
    stop_current();
    remove(g_binary);
}

/* ── main ─────────────────────────────────────────────────────────────── */
int main(int argc, char *argv[])
{
    if (argc != 2) {
        fprintf(stderr, "usage: %s <file.c>\n", argv[0]);
        return 1;
    }

    /* resolve to absolute path without relying on realpath() */
    if (argv[1][0] == '/') {
        snprintf(g_src, sizeof(g_src), "%s", argv[1]);
    } else {
        char cwd[PATH_MAX];
        if (getcwd(cwd, sizeof(cwd)) == NULL) { perror("getcwd"); return 1; }
        snprintf(g_src, sizeof(g_src), "%s/%s", cwd, argv[1]);
    }
    /* quick existence check */
    struct stat _st;
    if (stat(g_src, &_st) != 0) { perror(g_src); return 1; }

    /* build binary path next to source: /path/to/.livec_out */
    char *slash = strrchr(g_src, '/');
    if (slash) {
        size_t dir_len = (size_t)(slash - g_src + 1);
        snprintf(g_binary, sizeof(g_binary), "%.*s.livec_out",
                 (int)dir_len, g_src);
    } else {
        snprintf(g_binary, sizeof(g_binary), ".livec_out");
    }

    struct sigaction sa = { .sa_handler = on_sigint };
    sigaction(SIGINT, &sa, NULL);

    watch();
    return 0;
}
